clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/InactiveTopicPartitionChecker.java

```diff
@@ -0,0 +1,221 @@
+ *   <li>Pause inactive topic partitions to prevent them from blocking consumption of active topic partitions</li>
+ *   <li>Resume previously paused topic partitions that have become active again</li>
+ * </ul>
+ *
+ * <p>The service maintains state about which topic partitions are currently paused for each consumer
+ * and makes intelligent decisions about when to pause/resume based on the current activity status.
+ *
+ * <p>This is particularly useful in scenarios where some topic partitions may become temporarily
+ * unavailable or slow, preventing the consumer from making progress on other healthy topic partitions.
+ *
+ * <p>The checker uses per-topic-partition statistics from {@link ConsumptionTask} to determine activity,
+ * specifically looking at the last successful poll timestamp for each topic partition.
+ */
+public class InactiveTopicPartitionChecker extends AbstractVeniceService {
+  private static final Logger LOGGER = LogManager.getLogger(InactiveTopicPartitionChecker.class);
+
+  /** Map tracking which topic partitions are currently paused for each consumer */
+  private final Map<SharedKafkaConsumer, Set<PubSubTopicPartition>> consumerToPausedTopicPartitionsMap =
+      new VeniceConcurrentHashMap<>();
+
+  /** Map of consumers to their associated consumption tasks */
+  private final IndexedMap<SharedKafkaConsumer, ConsumptionTask> consumerToConsumptionTask;
+
+  /** Interval in milliseconds between inactive topic partition checks */
+  private final long inactiveTopicPartitionCheckIntervalInMs;
+
+  /** Threshold in milliseconds after which a topic partition is considered inactive */
+  private final long inactiveTopicPartitionThresholdInMs;
+
+  /** Executor service for scheduling periodic inactive topic partition checks */
+  private final ScheduledExecutorService executorService;
+
+  /**
+   * Creates a new InactiveTopicPartitionChecker.
+   *
+   * @param consumerToConsumptionTask map of consumers to their consumption tasks,
+   *                                  used to check topic partition activity status
+   * @param inactiveTopicPartitionCheckIntervalInSeconds how often to check for inactive topic partitions (in milliseconds)
+   * @param inactiveTopicPartitionThresholdInSeconds threshold after which a topic partition is considered inactive (in milliseconds)
+   */
+  public InactiveTopicPartitionChecker(
+      IndexedMap<SharedKafkaConsumer, ConsumptionTask> consumerToConsumptionTask,
+      long inactiveTopicPartitionCheckIntervalInSeconds,
+      long inactiveTopicPartitionThresholdInSeconds) {
+    this.consumerToConsumptionTask = consumerToConsumptionTask;
+    this.inactiveTopicPartitionCheckIntervalInMs =
+        TimeUnit.SECONDS.toMillis(inactiveTopicPartitionCheckIntervalInSeconds);
+    this.inactiveTopicPartitionThresholdInMs = TimeUnit.SECONDS.toMillis(inactiveTopicPartitionThresholdInSeconds);
+    this.executorService = Executors.newSingleThreadScheduledExecutor();
```

Comment:
Can we use daemon thread factory here and also pass log context?
```java
Executors.newSingleThreadScheduledExecutor(new DaemonThreadFactory("ConsumptionMonitor", logContext);
```

Reply:
Sure, I can add the DaemonThreadFactory but we don't have the log context here handy so I will just not doing that.